Palace/Accounts/Library/AccountsManager.swift

```diff
@@ -40,11 +40,23 @@ let currentAccountIdentifierKey  = "TPPCurrentAccountIdentifier"
   private var accountSet: String
 
   private var accountSets = [String: [Account]]()
-  private var accountSetsWorkQueue = DispatchQueue(label: "org.thepalaceproject.palace.AccountsManager.workQueue", attributes: .concurrent)
+  
+  private let accountSetsLock = NSRecursiveLock()
+  
+  /// Performs a closure within a lock using `accountSetsLock`
+  /// - Parameter action: the action inside the locked
+  private func performLocked(_ action: () -> Void) {
+    accountSetsLock.lock()
+    defer {
+      accountSetsLock.unlock()
+    }
+    action()
+  }
 
   var accountsHaveLoaded: Bool {
     var accounts: [Account]?
-    accountSetsWorkQueue.sync {
```

Comment:
Why is this only now causing issues?

Reply:
We didn't have this problem before because this is new code in TPPNetworkExecutor.swift.

1. We check the type of current account before calling `executeRequest` in TPPNetworkExecutor.swift, line 84
`if let authDefinition = TPPUserAccount.sharedAccount().authDefinition ...`, 
(I added this line to avoid username and password auth for SAML accounts),
- that calls `account(_ uuid: String) -> Account?` inside, in AccountsManager.swift, line 282,
- that contains that `.sync { ... }` block.

2. Because this is a part of the `executeRequest` function, it's also called inside `Account.init()`, when it downloads account logo during initialization.

3. And it happens so that we call the Account initializer within another `.sync { ... }` block in AccountsManager.swift in `private func loadAccountSetsAndAuthDoc(fromCatalogData data: Data, key: String, completion: @escaping (Bool) -> ()) {`.

It would be possible to solve that by moving `catalogsFeed.catalogs.map { Account(publication: $0) }` outside the `.sync { ... }` block (which I did), but then I also removed .sync { ... } blocks as they are not really safe.

Reply:
Ah I see, thanks for the explanation.